### ISSUE TITLE
Store immutability

### ISSUE THREAD

**Issue by jareware** [goatslacker/alt#19]
Alt looks like it's got a lot of good ideas on improving Flux.

The README talks about store immutability, but I don't see much pertaining to that in the code, nor do I see dependencies on e.g. Immutable.js or `mori`.

Could you elaborate a bit on this?

**Comment by cdebotton** [comment]
To add to this, I've tried using Immutable.js in an alt store and am having difficulty working with an isomorphic server. My problem is that when the server initially loads, the Immutable data will exist as the proper reference to the Immutable Map. On every subsequent page load, the immutable data is converted to a standard JavaScript object.

I have the following code in a store:

```js
  constructor() {
    this.bindActions(UserActionCreators);

    this.users = {};

    this.test = Immutable.Map({
      a: 'foo',
      b: 'bar'
    });

    this.on('snapshot', () => {
      if (Immutable.Map.isMap(this.test)) {
        this.test = this.test.toJS();
      }
    });

    this.on('bootstrap', () => {
      if (! Immutable.Map.isMap(this.test)) {
        this.test = Immutable.fromJS(this.test);
      }
    });
  }
```

Any guidance on how to handle this would be greatly appreciated.

**Comment by goatslacker** [comment]
@jareware I've updated the README to better clarify what is going on: 44effb026482fbf004f9d18f882a6f43cc46d5c1

TL;DR `getState` shallow copies your store's state so you don't accidentally overwrite it.

@cdebotton you may need to add another method. 

I'm assuming that once you bootstrap you use `flush()` and if that is the case then stores are recycled using `recycle()` which means you need to hook into the init lifecycle.

```js
 this.on('init', () => {
      if (! Immutable.Map.isMap(this.test)) {
        this.test = Immutable.fromJS(this.test);
      }
  });
```

I can actually create a built-in way of making this work with immutable.js a bit easier if you can give me more context.

**Comment by jareware** [comment]
Thanks, that clarifies it for me!

Consider the issue resolved on my part.

**Comment by cdebotton** [comment]
@goatslacker, thanks for pointing me in the right direction. I got this working by writing a bootstrap method that both init and bootstrap's life cycle events called. From what I can tell, in order for both the server and client environments to re-encapsulate the snapshot in Immutable, 'init' needs to be run on the server and 'bootstrap' needs to be run on the client.

```js
  constructor() {
    this.on('init', this.bootstrap);
    this.on('bootstrap', this.bootstrap);

    this.bindActions(UserActionCreators);

    this.test = Map({ a: 'Foo', b: 'bar' });
  }

  bootstrap() {
    if (! Map.isMap(this.test)) {
      this.test = fromJS(this.test);
    }
  }
```
